File: tools/extract_food_table.js
const fs = require('fs');

const foodFile = 'e:\\Steam\\steamapps\\common\\ProjectZomboid\\media\\scripts\\generated\\items\\food.txt';
const outDir = 'd:\\project\\pzmod\\docs\\superpowers\\specs\\current';

const content = fs.readFileSync(foodFile, 'utf-8');

function parseItems(text) {
    const items = [];
    const regex = /item\s+(\w+)\s*\{/g;
    let m;
    while ((m = regex.exec(text)) !== null) {
        const name = m[1];
        const start = m.index + m[0].length;
        let depth = 1;
        let j = start;
        while (j < text.length && depth > 0) {
            if (text[j] === '{') depth++;
            else if (text[j] === '}') depth--;
            j++;
        }
        const block = text.substring(start, j - 1);
        items.push({ name, block });
    }
    return items;
}

function extractField(block, field) {
    const re = new RegExp(field + '\\s*=\\s*([^,\\n]+)');
    const m = block.match(re);
    return m ? m[1].trim() : null;
}

const items = parseItems(content);
const foods = [];

for (const { name, block } of items) {
    const itemType = extractField(block, 'ItemType');
    if (itemType && itemType.toLowerCase().includes('food')) {
        foods.push({
            id: 'Base.' + name,
            weight: extractField(block, 'Weight'),
            hunger: extractField(block, 'HungerChange'),
            thirst: extractField(block, 'ThirstChange'),
            calories: extractField(block, 'Calories'),
            carbs: extractField(block, 'Carbohydrates'),
            lipids: extractField(block, 'Lipids'),
            proteins: extractField(block, 'Proteins'),
            foodType: extractField(block, 'FoodType'),
            daysFresh: extractField(block, 'DaysFresh'),
            daysRotten: extractField(block, 'DaysTotallyRotten'),
            cantEat: extractField(block, 'CantEat'),
            eatType: extractField(block, 'EatType'),
        });
    }
}

foods.sort((a, b) => {
    const fa = a.foodType || 'Unknown';
    const fb = b.foodType || 'Unknown';
    if (fa !== fb) return fa.localeCompare(fb);
    return a.id.localeCompare(b.id);
});

const groups = {};
for (const f of foods) {
    const ft = f.foodType || 'Unknown';
    if (!groups[ft]) groups[ft] = [];
    groups[ft].push(f);
}

let md = '# PZ B42 食物数据表（从游戏源码提取）\n\n';
md += `> 共 ${foods.length} 种食物，来源：media/scripts/items/food.txt\n\n`;

md += '## 字段说明\n\n';
md += '| 字段 | 含义 |\n|------|------|\n';
md += '| id | 物品 ID（Base.xxx），用于 AddItem |\n';
md += '| weight | 重量 kg |\n';
md += '| hunger | HungerChange（负=解饿） |\n';
md += '| thirst | ThirstChange（负=解渴） |\n';
md += '| calories | 热量 |\n';
md += '| carbs | 碳水化合物 |\n';
md += '| lipids | 脂肪 |\n';
md += '| proteins | 蛋白质 |\n';
md += '| foodType | 食物类别 |\n';
md += '| daysFresh | 新鲜天数 |\n';
md += '| daysRotten | 完全腐烂天数 |\n';
md += '| cantEat | true=需开罐/加工后才能吃 |\n';
md += '| eatType | 进食方式 |\n\n';

for (const ft of Object.keys(groups).sort()) {
    const items = groups[ft];
    md += `## ${ft}（${items.length} 种）\n\n`;
    md += '| id | weight | hunger | thirst | calories | carbs | lipids | proteins | daysFresh | daysRotten | cantEat | eatType |\n';
    md += '|----|--------|--------|--------|----------|-------|--------|----------|-----------|------------|---------|---------|\n';
    for (const f of items) {
        md += `| ${f.id} | ${f.weight || '-'} | ${f.hunger || '-'} | ${f.thirst || '-'} | ${f.calories || '-'} | ${f.carbs || '-'} | ${f.lipids || '-'} | ${f.proteins || '-'} | ${f.daysFresh || '-'} | ${f.daysRotten || '-'} | ${f.cantEat || '-'} | ${f.eatType || '-'} |\n`;
    }
    md += '\n';
}

const outPath = outDir + '\\pz-b42-food-table.md';
fs.writeFileSync(outPath, md, 'utf-8');

console.log(`提取完成：${foods.length} 种食物，分组 ${Object.keys(groups).length} 类`);
console.log(`输出：${outPath}`);
console.log('');
console.log('=== 含水食物（thirst < 0）===');
for (const f of foods) {
    if (f.thirst) {
        const t = parseFloat(f.thirst);
        if (!isNaN(t) && t < 0) {
            console.log(`  ${f.id}: thirst=${f.thirst}, weight=${f.weight}`);
        }
    }
}
console.log('');
console.log('=== 常用测试食物 ===');
const testIds = ['Base.Apple', 'Base.Bread', 'Base.TinnedSoupOpen', 'Base.CannedMushroomSoupOpen',
    'Base.Watermelon', 'Base.Carrot', 'Base.Potato', 'Base.CannedCornOpen'];
for (const f of foods) {
    if (testIds.includes(f.id)) {
        console.log(`  ${f.id}: weight=${f.weight}, hunger=${f.hunger}, thirst=${f.thirst}, cal=${f.calories}, carbs=${f.carbs}`);
    }
}

File: tools/extract_food_table.py
import re, os, json

food_file = r'e:\Steam\steamapps\common\ProjectZomboid\media\scripts\generated\items\food.txt'
out_dir = r'd:\project\pzmod\docs\superpowers\specs\current'

with open(food_file, 'r', encoding='utf-8', errors='ignore') as f:
    content = f.read()

# 匹配 item 块（处理嵌套大括号）
def parse_items(text):
    items = []
    i = 0
    while True:
        m = re.search(r'item\s+(\w+)\s*\{', text[i:])
        if not m:
            break
        name = m.group(1)
        start = i + m.end()
        # 找到匹配的 closing brace
        depth = 1
        j = start
        while j < len(text) and depth > 0:
            if text[j] == '{':
                depth += 1
            elif text[j] == '}':
                depth -= 1
            j += 1
        block = text[start:j-1]
        items.append((name, block))
        i = j
    return items

items = parse_items(content)

def extract_field(block, field):
    m = re.search(field + r'\s*=\s*([^,\n]+)', block)
    return m.group(1).strip() if m else None

foods = []
for name, block in items:
    item_type = extract_field(block, 'ItemType')
    if item_type and 'food' in item_type.lower():
        food = {
            'id': 'Base.' + name,
            'weight': extract_field(block, 'Weight'),
            'hunger': extract_field(block, 'HungerChange'),
            'thirst': extract_field(block, 'ThirstChange'),
            'calories': extract_field(block, 'Calories'),
            'carbs': extract_field(block, 'Carbohydrates'),
            'lipids': extract_field(block, 'Lipids'),
            'proteins': extract_field(block, 'Proteins'),
            'foodType': extract_field(block, 'FoodType'),
            'daysFresh': extract_field(block, 'DaysFresh'),
            'daysRotten': extract_field(block, 'DaysTotallyRotten'),
            'cantEat': extract_field(block, 'CantEat'),
            'eatType': extract_field(block, 'EatType'),
        }
        foods.append(food)

foods.sort(key=lambda x: (str(x['foodType']), x['id']))

groups = {}
for f in foods:
    ft = f['foodType'] or 'Unknown'
    groups.setdefault(ft, []).append(f)

md_lines = ['# PZ B42 食物数据表（从游戏源码提取）', '', f'> 共 {len(foods)} 种食物，来源：media/scripts/items/food.txt', '']

md_lines.append('## 字段说明')
md_lines.append('| 字段 | 含义 |')
md_lines.append('|------|------|')
md_lines.append('| id | 物品 ID（Base.xxx），用于 AddItem |')
md_lines.append('| weight | 重量 kg |')
md_lines.append('| hunger | HungerChange（负=解饿） |')
md_lines.append('| thirst | ThirstChange（负=解渴） |')
md_lines.append('| calories | 热量 |')
md_lines.append('| carbs | 碳水化合物 |')
md_lines.append('| lipids | 脂肪 |')
md_lines.append('| proteins | 蛋白质 |')
md_lines.append('| foodType | 食物类别 |')
md_lines.append('| daysFresh | 新鲜天数 |')
md_lines.append('| daysRotten | 完全腐烂天数 |')
md_lines.append('| cantEat | true=需开罐/加工后才能吃 |')
md_lines.append('| eatType | 进食方式 |')
md_lines.append('')

for ft, items in sorted(groups.items()):
    md_lines.append(f'## {ft}（{len(items)} 种）')
    md_lines.append('')
    md_lines.append('| id | weight | hunger | thirst | calories | carbs | lipids | proteins | daysFresh | daysRotten | cantEat | eatType |')
    md_lines.append('|----|--------|--------|--------|----------|-------|--------|----------|-----------|------------|---------|---------|')
    for f in items:
        row = '| ' + ' | '.join([
            f['id'],
            f['weight'] or '-',
            f['hunger'] or '-',
            f['thirst'] or '-',
            f['calories'] or '-',
            f['carbs'] or '-',
            f['lipids'] or '-',
            f['proteins'] or '-',
            f['daysFresh'] or '-',
            f['daysRotten'] or '-',
            f['cantEat'] or '-',
            f['eatType'] or '-',
        ]) + ' |'
        md_lines.append(row)
    md_lines.append('')

out_path = os.path.join(out_dir, 'pz-b42-food-table.md')
with open(out_path, 'w', encoding='utf-8') as f:
    f.write('\n'.join(md_lines))

print(f'提取完成：{len(foods)} 种食物，分组 {len(groups)} 类')
print(f'输出：{out_path}')
print()
print('=== 含水食物（thirst < 0）===')
for f in foods:
    if f['thirst']:
        try:
            if float(f['thirst']) < 0:
                print(f"  {f['id']}: thirst={f['thirst']}, weight={f['weight']}")
        except:
            pass

print()
print('=== 常用测试食物 ===')
test_ids = ['Base.Apple', 'Base.Bread', 'Base.TinnedSoupOpen', 'Base.CannedMushroomSoupOpen',
            'Base.Watermelon', 'Base.Carrot', 'Base.Potato', 'Base.CannedCornOpen']
for f in foods:
    if f['id'] in test_ids:
        print(f"  {f['id']}: weight={f['weight']}, hunger={f['hunger']}, thirst={f['thirst']}, cal={f['calories']}, carbs={f['carbs']}")
